executor/common_kvm_arm64_syzos.h

```diff
@@ -185,3 +203,434 @@ GUEST_CODE static void guest_handle_hvc(struct api_call_smccc* cmd)
+	spin_while_readl(ARM64_ADDR_GICR_BASE + GICR_WAKER, GICR_WAKER_ChildrenAsleep);
+}
+
+GUEST_CODE void gicv3_cpu_init(uint32 cpu)
+{
+	uint64 sgi_base = sgi_base_cpu(cpu);
+
+	// It is important that software performs these steps before configuring
+	// the CPU interface, otherwise behavior can be UNPREDICTABLE.
+	gicv3_enable_redist(cpu);
+
+	// Mark all the SGI and PPI interrupts as non-secure Group-1. Also, deactivate and disable them.
+	writel(~0, sgi_base + GICR_IGROUPR0);
+	writel(~0, sgi_base + GICR_ICACTIVER0);
+	writel(~0, sgi_base + GICR_ICENABLER0);
+
+	// Set a default priority for all the SGIs and PPIs.
+	for (int i = 0; i < 32; i += 4) {
+		writel(GICD_INT_DEF_PRI_X4,
+		       sgi_base + GICR_IPRIORITYR0 + i);
+	}
+
+	gicr_wait_for_rwp(cpu);
+
+	// Enable the GIC system register (ICC_*) access.
+	uint32 icc_sre_el1 = 0;
+	asm volatile("mrs %0, " ICC_SRE_EL1
+		     :
+		     : "r"(icc_sre_el1));
+	icc_sre_el1 |= ICC_SRE_EL1_SRE;
+	asm volatile("msr " ICC_SRE_EL1 ", %0"
+		     :
+		     : "r"(icc_sre_el1));
+
+	// Set a default priority threshold.
+	uint32 value = ICC_PMR_DEF_PRIO;
+	asm volatile("msr " ICC_PMR_EL1 ", %0"
+		     :
+		     : "r"(value));
+
+	// Enable non-secure Group-1 interrupts.
+	value = ICC_IGRPEN1_EL1_ENABLE;
+	asm volatile("msr " ICC_IGRPEN1_EL1 ", %0"
+		     :
+		     : "r"(value));
+}
+
+// GICv3 reserves interrupts 32-1019 for SPI.
+#define VGICV3_MIN_SPI 32
+#define VGICV3_MAX_SPI 1019
+
+// https://developer.arm.com/documentation/ihi0048/b/Programmers--Model/Distributor-register-descriptions/Interrupt-Set-Enable-Registers--GICD-ISENABLERn
+GUEST_CODE void gicv3_irq_enable(uint32 intid)
+{
+	// TODO(glider): support multiple CPUs. E.g. KVM selftests store CPU ID in TPIDR_EL1.
+	uint32 cpu = 0;
+
+	writel(1 << (intid % 32), ARM64_ADDR_GICD_BASE + GICD_ISENABLER + (intid / 32) * 4);
+	if ((intid >= VGICV3_MIN_SPI) && (intid <= VGICV3_MAX_SPI))
+		gicd_wait_for_rwp();
+	else
+		gicr_wait_for_rwp(cpu);
+}
+
+GUEST_CODE static void guest_handle_irq_setup(struct api_call_irq_setup* cmd)
+{
+	int nr_spis = cmd->nr_spis;
+	if ((nr_spis > VGICV3_MAX_SPI - VGICV3_MIN_SPI) || (nr_spis < 0))
+		nr_spis = 32;
+	int nr_cpus = cmd->nr_cpus;
+
+	gicv3_dist_init(nr_spis);
+	for (int i = 0; i < nr_cpus; i++)
+		gicv3_cpu_init(i);
+	for (int i = 0; i < nr_spis; i++)
+		gicv3_irq_enable(VGICV3_MIN_SPI + i);
+	// Set up the vector table.
+	asm("adr x1, guest_vector_table\n");
+	asm("msr vbar_el1, x1\n");
+	asm("msr daifclr, #0b1111\n");
+}
+
+// Registers saved by one_irq_handler() and received by guest_irq_handler().
+struct ex_regs {
+	uint64 regs[31];
+	uint64 sp;
+	uint64 pc;
+	uint64 pstate;
+};
+
+// Placeholder function to declare one_irq_handler() inside the assembly blob. We cannot put it
+// into a separate .S file, because syzkaller requires a standalone header for reproducers.
+__attribute__((used))
+GUEST_CODE static void
+one_irq_handler_fn()
+{
+	asm volatile(
+	    ".global one_irq_handler\n"
```

Comment:
I find large asm blocks more readable with raw string literal, then you don't need `"` and clutter the code with `\n`:
```
	asm(R"(
        ...
	)");
```